# EDA (exploratory data analysis): Diagnosing Diabetes

Exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

Using EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

In [2]:
#2 load in data
diabetes_data = pd.read_csv('diabetes.csv', sep=',', encoding='utf-8')

In [3]:
# print first 10 rows 
diabetes_data.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1  
5                     0.201   30       0  
6                     0.248   26       1  
7                     0.134   29       0  
8                     0.158   53       1  
9                     0.232   54       1

In [4]:
# the numbers of rows and columns by method .shape
diabetes_data.shape

(768, 9)

In [5]:
# which columns are there
diabetes_data.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [7]:
# check columns contain missing data
print(diabetes_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(2), int64(6), object(1)
memory usage: 54.1+ KB
None


In [8]:
# another check columns contain missing data
print(diabetes_data.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


In [9]:
# calculate summary statistics on diabetes_data using the .describe() method
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000                  0.243750   24.000000  
50%     32.000000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000

All column values at the row 'min' are 0.000000

The maximum value of the Insulin column is 846, which is abnormally high.
The maximum value of the Pregnancies column is 17.

Need to look further.

In [10]:
# replace the instances of 0 with NaN in the five columns
diabetes_data[['Pregnancies', 
               'Glucose', 
               'BloodPressure', 
               'SkinThickness', 
               'Insulin',
               'BMI']] = diabetes_data[['Pregnancies', 
                                        'Glucose', 
                                        'BloodPressure', 
                                        'SkinThickness', 
                                        'Insulin', 
                                        'BMI']].replace(0, np.nan)

In [11]:
# one more time check columns contain missing data by 2 methods:
print(diabetes_data.isnull().sum())
print()
print(diabetes_data.info())

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               657 non-null    float64
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null 

In [12]:
# print out all of the rows that contain missing (null) values
diabetes_data[diabetes_data.isnull().any(axis=1)].head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0           6.0    148.0           72.0           35.0      NaN  33.6   
1           1.0     85.0           66.0           29.0      NaN  26.6   
2           8.0    183.0           64.0            NaN      NaN  23.3   
4           NaN    137.0           40.0           35.0    168.0  43.1   
5           5.0    116.0           74.0            NaN      NaN  25.6   
7          10.0    115.0            NaN            NaN      NaN  35.3   
9           8.0    125.0           96.0            NaN      NaN   NaN   
10          4.0    110.0           92.0            NaN      NaN  37.6   
11         10.0    168.0           74.0            NaN      NaN  38.0   
12         10.0    139.0           80.0            NaN      NaN  27.1   

    DiabetesPedigreeFunction  Age Outcome  
0                      0.627   50       1  
1                      0.351   31       0  
2                      0.672   32       1  
4                      2.288   33       1  
5                      0.201   30       0  
7                      0.134   29       0  
9                      0.232   54       1  
10                     0.191   30       0  
11                     0.537   34       1  
12                     1.441   57       0

Most rows with missing data have missing values in more than one column.  Every single row with at least one missing value also has a missing value in the 'insulin' column.

In [13]:
# check columns data types by method .dtypes
diabetes_data.dtypes

Pregnancies                 float64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                      object
dtype: object

'Outcome' column is of type object. Let’s figure out why this might be.

In [15]:
# print out unique values of Outcome column
print(diabetes_data['Outcome'].unique())

['1' '0' 'O']


We have instances of the character 'O' in addition to the number 0

In [18]:
# replace instances of 'O' with 0 and convert the Outcome column to type int64
diabetes_data['Outcome'] = diabetes_data['Outcome'].replace('O', 0).astype("int64")

In [19]:
print(diabetes_data['Outcome'].unique())

[1 0]


In [21]:
# acheck columns data types by method .dtypes
diabetes_data.dtypes

Pregnancies                 float64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [23]:
# using .value_count method to show count of pregnancies
print(diabetes_data['Pregnancies'].value_counts())

Pregnancies
1.0     135
2.0     103
3.0      75
4.0      68
5.0      57
6.0      50
7.0      45
8.0      38
9.0      28
10.0     24
11.0     11
13.0     10
12.0      9
14.0      2
15.0      1
17.0      1
Name: count, dtype: int64


In [24]:
# using .value_count method to show count of ages
print(diabetes_data['Age'].value_counts())

Age
22    72
21    63
25    48
24    46
23    38
28    35
26    33
27    32
29    29
31    24
41    22
30    21
37    19
42    18
33    17
38    16
36    16
32    16
45    15
34    14
46    13
43    13
40    13
39    12
35    10
50     8
51     8
52     8
44     8
58     7
47     6
54     6
49     5
48     5
57     5
53     5
60     5
66     4
63     4
62     4
55     4
67     3
56     3
59     3
65     3
69     2
61     2
72     1
81     1
64     1
70     1
68     1
Name: count, dtype: int64
